crates/s3s-fs/src/main.rs

```diff
@@ -108,18 +111,36 @@ async fn run(opt: Opt) -> Result {
     };
 
     // Run server
-    let listener = TcpListener::bind((opt.host.as_str(), opt.port))?;
+    let listener = TcpListener::bind((opt.host.as_str(), opt.port)).await?;
     let local_addr = listener.local_addr()?;
 
-    let server = Server::from_tcp(listener)?.serve(service.into_shared().into_make_service());
+    let hyper_service = service.into_shared();
+
+    let connection = ConnBuilder::new(TokioExecutor::new());
+
+    let server = async move {
+        loop {
+            let (socket, _) = match listener.accept().await {
+                Ok(ok) => ok,
+                Err(err) => {
+                    tracing::error!("error accepting connection: {err}");
+                    continue;
+                }
+            };
+            let service = hyper_service.clone();
+            let conn = connection.clone();
+            tokio::spawn(async move {
+                let _ = conn.serve_connection(TokioIo::new(socket), service).await;
+            });
+        }
+    };
 
+    let task = tokio::spawn(server);
     info!("server is running at http://{local_addr}");
-    server.with_graceful_shutdown(shutdown_signal()).await?;
+
+    tokio::signal::ctrl_c().await?;
+    task.abort();
```

Comment:
We are waiting for graceful shutdown utility in hyper-util. https://github.com/hyperium/hyper-util/pull/86
Aborting the service may cause problems because s3s-fs doesn't have a great design for concurrent security.

Reply:
ok the new PR is https://github.com/hyperium/hyper-util/pull/108

Reply:
It seems that they still need some time. We can move on without this utility.